Reenable RequestStream tests and address HttpWebRequest PR feedback

ViktorHofer opened the pull request:
Found the root cause why CI was hanging yesterday. `RequestStreamTest` class fires ~20 different tests in parallel. They all have something in common, they call `HttpWebRequest.GetRequestStream()`. As the limit in netfx (`ServicePointManager.DefaultConnectionLimit`) is very low, the test execution timed out.

Related: https://github.com/dotnet/corefx/pull/19584
See here: https://github.com/dotnet/corefx/issues/16928

ViktorHofer commented:
@dotnet-bot test Outerloop netfx Windows_NT Release x64 Build and Test
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64

ViktorHofer commented:
@stephentoub I added the removal of the dispose what you mentioned in the other PR

stephentoub commented:
Thanks

ViktorHofer commented:
@dotnet-bot test Outerloop netfx Windows_NT Release x64 Build and Test
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64 Build and Test

safern commented on src/System.Net.Requests/tests/RequestStreamTest.cs:
```diff
@@ -11,11 +11,18 @@
+            // On .NET Framework, the default limit for connections/server is very low (2).
+            // On .NET Core, the default limit is higher. Since these tests run in parallel,
+            // the limit needs to be increased to avoid timeouts when running the tests.
+            ServicePointManager.DefaultConnectionLimit = int.MaxValue;
```
@stephentoub so this was the reason why we where getting hang on this test class for desktop: https://github.com/dotnet/corefx/issues/16928

You okay with this change?

stephentoub replied:
So basically the tests are creating a leak, and rather than fixing the leak we're increasing the size of the bucket?  It'd be better to just fix the leak, e.g. if the problem is that GetRequestStream is leaving open a connection because we haven't actually asked for the response, we should ask for the response and dispose of it.

davidsh replied:
I don't think this is about a leak.  It is more about parallel requests especially true given how Xunit works in these libraries.  Since everything is using either the Azure echo server or the loopback server, that means a single ServicePoint object is being created and used.  And therefore, there will be a limit on how many connections can be open (in parallel) at once.

ViktorHofer replied:
I just tested what @stephentoub said and it indeed works. As soon as we ask for the response and dispose it, the tests complete.

ViktorHofer replied:
@davidsh see here: https://github.com/dotnet/corefx/issues/16928#issuecomment-288827780

davidsh replied:
>As soon as we ask for the response and dispose it, the tests complete.

This isn't my favorite way of solving this problem.  The tests here are about testing the request stream.  We're adding complexity to the tests by asking for the response as well. So, in terms of isolating what we should be unit-testing, adding complexity is not my favorite solution.

safern replied:
This is the bug in desktop we had the discussion about like a month ago when we opened #16928. This is not a leak, it is because in Desktop we open a connection to the server but in Core buffer the content as if in desktop .AllowWriteStreamBuffering was set to true. (https://github.com/dotnet/corefx/issues/11873#issuecomment-248124938). So maybe another workaround will be setting .AllowWriteStreamBuffering to true and disposing the WebRequest so that it is actually written to the server?

stephentoub replied:
It is a leak: every time we call GetRequestStream, we're leaking a connection that remains open... that's why we hit the limit and deadlock forever rather than it just throttling how fast the tests run.

davidsh replied:
Thanks.  I remember the conversation.

So, it's still not a leak, per se, but just part of the way that networking works.  It's expected that if you ask for a request stream is network bound....you get an open connection until you close it / dispose it / finalize it.

So, the choice here is a trade-off.  Add test complexity by asking for a response which does more network I/o (and might mask what we're trying to test with the request stream)....or alter the connection pool to be able to handle the parallel tests.

safern replied:
So talked offline to @ViktorHofer and yes there is a leak because we never dispose the RequestStream, so we should then dispose the GetRequestStream so that the connection is closed.

safern replied:
Missed completely that we where not disposing them. Thanks @stephentoub for explaning and pointing it out.

HttpWebRequestTests have the same "fix" 
https://github.com/dotnet/corefx/blob/master/src/System.Net.Requests/tests/HttpWebRequestTest.cs#L34

So should we fix it on that side as well?

ViktorHofer replied:
What makes me wonder, why isn't disposing the RequestStream closing the connection?

stephentoub replied:
> because we never dispose the RequestStream

We do Dispose of the request stream: they're all (or at least the ones I looked at) wrapped in using blocks.  The problem is that there appears to be a bug in desktop where disposing of the request stream is insufficient to actually release the connection, and to do that you need to actually get the response so you can dispose of it.

I'm not sure how disposing it masks what we're trying to do with the request stream: we just create a helper and change a test like:
```C#
using (Stream stream = GetRequestStream())
{
    Assert.Throws<ArgumentNullException>("buffer", () => stream.Write(null, 0, 1));
}
```
to something like:
```C#
using (RequestStreamWrapper w = GetRequestStream())
{
    Assert.Throws<ArgumentNullException>("buffer", () => w.Stream.Write(null, 0, 1));
}
```
where the RequestStreamWrapper stores both the Stream and the HttpWebRequest, such that its Dispose (called implicitly by the using) can do GetResponse().Dispose on the request... it could even choose only to do this for desktop where there's the bug we're working around.

ViktorHofer replied:
> The problem is that there appears to be a bug in desktop where disposing of the request stream is insufficient to actually release the connection, and to do that you need to actually get the response so you can dispose of it.

I fully agree

davidsh replied:
I want to clarify that this is not a bug in Desktop.  This is part of the design of HttpWebRequest and GetRequestStream.

It should not close the TCP connection when you close/dispose/using the request stream.

The reason is that the semantics are that you are supposed to get the request stream, write to it, and then close/dispose it.  That completes the "request" portion of the HTTP request.  For cases where AllowWriteStreamBuffering is true, it will calculate the content size and adjust the Content-Length header.  For cases of chunked transfer, it will send the zero-length terminator.

But it should not close the TCP connection.  The connection has to stay open in order to be ready for someone to ask for the response via GetResponse.

davidsh replied:
cc: @CIPop @Tratcher

stephentoub replied:
Then if it's not a bug but instead part of the design, the tests should absolutely be following that design and not abandoning things halfway.  Not sure why you gave my comment a thumbs down.

ViktorHofer commented:
@davidsh @stephentoub Outerloop errors not realted to this PR:

```
System.Net.Tests.HttpListenerPrefixCollectionTests.Add_InvalidPrefixNotStarted_ThrowsHttpListenerExceptionOnStart(uriPrefix: \"http://43f9a5b55e494ca7973b5eebc494024d/\")
System.Net.Tests.HttpListenerPrefixCollectionTests.Add_InvalidPrefixAlreadyStarted_ThrowsHttpListenerExceptionOnAdd(uriPrefix: \"http://d260d73296f046459a0efa74613a56aa/\")
```

ViktorHofer commented:
@all PTAL. Should I do the same for HttpWebRequest in the same PR?

davidsh commented on src/System.Net.Requests/tests/HttpWebRequestTest.cs:
```diff
@@ -1161,21 +1076,6 @@ public void Method_SetInvalidString_ThrowsArgumentException(Uri remoteServer)
             Assert.Throws<ArgumentException>("value", () => request.Method = "Method(2");
         }
 
-        [Fact]
```
Why are we deleting all these tests in this file?

ViktorHofer replied:
Those were the tests I created with a previous PR. It made sense to combine them into one method so that instead of ~10 requests we only make one and test it in there. See L181

safern replied:
https://github.com/dotnet/corefx/pull/19613/files#diff-25f859e9dd82e6bb39d6bfc84f10fbe1R172

stephentoub commented on src/System.Net.Requests/tests/RequestStreamWrapper.cs:
```diff
@@ -0,0 +1,45 @@
+            }
+
+            // Disposing the request stream.
+            _stream.Dispose();
```
This should be done before the above block rather than after.

stephentoub commented on src/System.Net.Requests/tests/RequestStreamWrapper.cs:
```diff
@@ -0,0 +1,45 @@
+                    _stream = _request.GetRequestStream();
+                return _stream;
+            }
+        }
```
I don't think this needs to be done lazily: the function to get one of these is calling GetRequestStream... it can be prompt and call _request.GetRequestStream in the ctor.

stephentoub commented on src/System.Net.Requests/tests/RequestStreamWrapper.cs:
```diff
@@ -0,0 +1,45 @@
+        public RequestStreamWrapper(HttpWebRequest request)
+        {
+            if (request == null)
+                throw new ArgumentNullException(nameof(request));
```
Could just be an assert

Tratcher commented:
? DefaultConnectionLimit never applied to localhost.

ViktorHofer commented:
@Tratcher there are tests which uses a LoopbackServer in it but only a few. The others are testing against a remote server.

ViktorHofer commented:
@dotnet-bot test Outerloop netfx Windows_NT Release x64 Build and Test
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64 Build and Test

ViktorHofer commented:
@stephentoub sorry PTAL once more. Now everything should be covered.

ViktorHofer commented:
@dotnet-bot test Innerloop Tizen armel Debug Cross Build

stephentoub commented:
> DefaultConnectionLimit never applied to localhost.

This highlights another way we could fix this then: use a loopback server for these tests.

stephentoub commented on src/System.Net.Requests/tests/HttpWebRequestTest.cs:
```diff
@@ -930,33 +856,41 @@ public void BeginGetRequestStream_CreatePostRequestThenCallTwice_ThrowsInvalidOp
+            state.Request = request;
 
-            IAsyncResult asyncResult = request.BeginGetRequestStream(null, null);
+            IAsyncResult asyncResult = request.BeginGetRequestStream(RequestCallback, state);
```
This looks problematic: it's introducing a race condition, as the get request stream operation could complete before BeginGetRequestStream is called, at which point that should actually succeed rather than throw.

stephentoub commented on src/System.Net.Requests/tests/HttpWebRequestTest.cs:
```diff
@@ -930,33 +856,41 @@ public void BeginGetRequestStream_CreatePostRequestThenCallTwice_ThrowsInvalidOp
         }
 
         [Theory, MemberData(nameof(EchoServers))]
         [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "request stream not allowed for GET on netfx")]
```
Why do we allow it on corefx then?

davidsh replied:
Another instance of a behavior difference that was not planned.  I think I added that Skip* attribute.

Historically, we have not allowed GET w/ request-body for HttpWebRequest.  This is why GetRequestStream() would throw an exception on Desktop.   For HttpClient, we relaxed this restriction in the API surface.

We could decide to fix this in .NET Core and match the same exception behavior as Desktop.

stephentoub commented on src/System.Net.Requests/tests/HttpWebRequestTest.cs:
```diff
@@ -930,33 +856,41 @@ public void BeginGetRequestStream_CreatePostRequestThenCallTwice_ThrowsInvalidOp
+            state.Request = request;
 
-            IAsyncResult asyncResult = request.BeginGetResponse(null, null);
+            IAsyncResult asyncResult = request.BeginGetResponse(ResponseCallback, state);
```
Here as well I wouldn't rely on the callback, which will introduce races: I would instead synchronously block by explicitly calling EndGetResponse.

stephentoub commented on src/System.Net.Requests/tests/HttpWebRequestTest.cs:
```diff
@@ -1257,7 +1267,10 @@ public void HttpWebRequest_EndGetRequestStreamContext_ExpectedValue(Uri remoteSe
+                RequestState state = new RequestState();
+                state.Request = request;
+
+                request.BeginGetResponse(ResponseCallback, state);
```
This is changing what the test was trying to test... see the name of the method.

stephentoub replied:
(You don't need to pass callbacks in these cases... you can just take the resulting IAsyncResult that's returned and pass it to the corresponding End method.  That'll let you control exactly when the End is invoked.  And it'll ensure that exceptions are propagated out of there rather than being stored into state.SavedException, which it appears is being ignored in many of these cases.

ViktorHofer commented:
@stephentoub the last HttpWebRequests changes were problematic so I reverted those. Now I moved all members which used GetRequestStream without resolving it with GetResponse to LoopbackServer.

I am pretty happy now with the current state. Execution time is pretty fast (22sec on my machine) and most tests are now hardened.

PTAL

ViktorHofer commented:
@dotnet-bot test Outerloop netfx Windows_NT Release x64 Build and Test
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64 Build and Test

ViktorHofer commented:
@dotnet-bot test Outerloop netfx Windows_NT Release x64 Build and Test
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64 Build and Test

ViktorHofer commented:
@dotnet-bot test Innerloop netfx Windows_NT Release x64 Build and Test
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64

stephentoub commented:
I rebased to address the merge conflicts.

stephentoub commented:
@dotnet-bot test Innerloop Ubuntu14.04 Debug x64 Build and Test please
@dotnet-bot test Innerloop Ubuntu14.04 Release x64 Build and Test please
